### ISSUE TITLE
Jest - ExperienceForm axe tests too slow

### ISSUE THREAD

**Issue by petertgiles** [GCTC-NTGC/gc-digital-talent#4755]
The "foo type should have no accessibility errors" jest tests in frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.test.tsx often take more than the default 5000 ms timeout period.  A workaround was put in to increase the timeout to 10000 ms since it was causing Axe test failures (#4746).   Can we speed up these tests?

Acceptance Criteria:
- [ ] Manual setTimeout override removed
- [ ] Tests consistently complete in less than 5000 ms

**Comment by esizer** [comment]
I spent some time investigating this and took some steps to attempt to improve performance of these tests. Ultimately, these did not impact it, at all. From what I can tell, it seems to be an issue with cold start where the first test is taking ~60s and all subsequent ones take significantly less time (<1s).

## Changes made

- Added `--max-workers=2` flag to jest
- Replaced `ts-jest` with `@swc/jest` transformer (https://swc.rs/docs/usage/jest)
- Modified `axe-core` config to only test wcagaa and only report violations

## Notes

I should have saved the link but during my research, but essentially someone blamed this issue on `js-dom` being slow. They mentioned that when running `axe-core` within that context it could be very slow on complex components. This kind of makes sense to me since we only see the issue on our most complex components (`SearchContainer`, `ExperienceForm`).

## Possible Solution(?)

I'm wondering if we could test pieces of the larger component and make assumptions based on that then use `cypress-axe` to test larger pieces? What I'm thinking is:

### `jest-axe`

Only test small components (ideally only things that appear in `@common` and smaller building blocks in the projects.  If we were using the atomic pattern, we would run `jest-axe` on the atom and molecule equivalents.

### `cypress-axe`

This would be used to test how our smaller components interact with each other on a page. Back to the atomic pattern, these would be the organism equivalents. 

So, instead of testing the `ExperienceForm` for a11y violations using `jest-axe`, we would test the entire experience form page in `cypress-axe`.

> **Note**
> This only works if the post I read was correct about the performance issues being more related to axe running in `js-dom` so it may not fix the issue 😢

**Comment by patcon** [comment]
This seems like a long-short, but might it have something to do with us using radix-ui components, that make use of `React.creatPortal`?

I'm seeing something in the jest-axe docs about using `baseElement` instead of `container` when React Portals are being used, but we're definitely using `container` from the render function...

https://www.npmjs.com/package/jest-axe#:~:text=If%20you%27re%20using%20React%20Portals%2C%20use%20the%20baseElement%20instead%20of%20container%3A

**Comment by patcon** [comment]
Also, noting that @vd1992 saw these failures randomly resolving themselves yesterday:
https://talent-cloud.slack.com/archives/C0259G6KXJ6/p1674066634112719?thread_ts=1674066547.285989&cid=C0259G6KXJ6

**Comment by esizer** [comment]
> I'm seeing something in the jest-axe docs about using `baseElement` instead of `container` when React Portals are being used, but we're definitely using `container` from the render function...

Attempted this just now and it is leading to a **_very_** long list of false positives.

**Comment by esizer** [comment]
Since these tests have been passing lately, I did a check to see long they were taking and it seems they are running at a decent speed now 😕 They are taking anywhere from 7-17 seconds. Still kind of slow but nothing we should be super worried about, maybe?

## Recent Runs

```sh
@gc-digital-talent/web:test:   ExperienceForm
@gc-digital-talent/web:test:     ✓ award type should have no accessibility errors (17045 ms)
@gc-digital-talent/web:test:     ✓ community type should have no accessibility errors (13694 ms)
@gc-digital-talent/web:test:     ✓ education type should have no accessibility errors (11836 ms)
@gc-digital-talent/web:test:     ✓ personal type should have no accessibility errors (14327 ms)
@gc-digital-talent/web:test:     ✓ work type should have no accessibility errors (12204 ms)
```

```sh
@gc-digital-talent/web:test:   ExperienceForm
@gc-digital-talent/web:test:     ✓ award type should have no accessibility errors (10697 ms)
@gc-digital-talent/web:test:     ✓ community type should have no accessibility errors (8248 ms)
@gc-digital-talent/web:test:     ✓ education type should have no accessibility errors (8335 ms)
@gc-digital-talent/web:test:     ✓ personal type should have no accessibility errors (8244 ms)
@gc-digital-talent/web:test:     ✓ work type should have no accessibility errors (6939 ms)
```

```sh
@gc-digital-talent/web:test:   ExperienceForm
@gc-digital-talent/web:test:     ✓ award type should have no accessibility errors (12320 ms)
@gc-digital-talent/web:test:     ✓ community type should have no accessibility errors (10797 ms)
@gc-digital-talent/web:test:     ✓ education type should have no accessibility errors (9848 ms)
@gc-digital-talent/web:test:     ✓ personal type should have no accessibility errors (8998 ms)
@gc-digital-talent/web:test:     ✓ work type should have no accessibility errors (7171 ms)
```

```sh
@gc-digital-talent/web:test:   ExperienceForm
@gc-digital-talent/web:test:     ✓ award type should have no accessibility errors (10864 ms)
@gc-digital-talent/web:test:     ✓ community type should have no accessibility errors (9025 ms)
@gc-digital-talent/web:test:     ✓ education type should have no accessibility errors (8427 ms)
@gc-digital-talent/web:test:     ✓ personal type should have no accessibility errors (9320 ms)
@gc-digital-talent/web:test:     ✓ work type should have no accessibility errors (7270 ms)
```

**Comment by vd1992** [comment]
lets keep our fingers crossed since I don't think anyone is enthusiastic about trying to sort out jest/github performance quirks

**Comment by petertgiles** [comment]
I suggest leaving the issue open but maybe deprioritize it.  IMO over 10 seconds is still way too long.

**Comment by esizer** [comment]
> I suggest leaving the issue open but maybe deprioritize it. IMO over 10 seconds is still way too long.

I agree, though I'm not sure we can drive it much lower with these forms. The axe tests just take that long  to run (especially with how large the `ExperienceForm` is) which is part of the reason I suggest maybe being more selective about what components we do the a11y tests for in jest. 

Since in the experience form all we are really testing here is:

 - Some inputs
 - Skill Picker
 - Headings

I think if we do axe tests for each input component in `common/.../forms`, then run a cypress-axe test on the page itself we cover our bases without bogging down jest with overly complex components.

This is quite an old comment but I think it describes our situation well.

https://github.com/dequelabs/axe-core/issues/325#issuecomment-299244385

**Comment by petertgiles** [comment]
Yeah, fair enough.  Maybe this approach is just wrong for a big form.  A solution for Axejest definitely isn't presenting itself.  :laughing:

**Comment by petertgiles** [comment]
I thought of another reason to stop Axe testing in Jest - there are no Hydrogen styles there so it's probably not testing a very high fidelity version of the components.
Ref: #2080

**Comment by vd1992** [comment]
Is this issue still relevant? Especially if we may want to switch away from testing axe in jest, in favour of Storybook maybe. 
The tests aren't terribly slow

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/26997bb5-b54e-4ff3-9f36-bdb95897d4bf)

Though, if you want to optimize tests there is another very good candidate around. 

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/17a2573f-c57c-4605-8764-7d1a381e0822)

And scrum boss, @mnigh

**Comment by mnigh** [comment]
> Is this issue still relevant? Especially if we may want to switch away from testing axe in jest, in favour of Storybook maybe. The tests aren't terribly slow

There has been a lot of discussion on this issue since it was estimated. I think it is worth talking about again as devs and refining it more.

**Comment by mnigh** [comment]
Tests are not as slow anymore.
